### PR TITLE
Add visualization

### LINKED ISSUE #64: Renaming args in `clinicadl train`
Some arguments could be changed:
- `use_extracted_patches|roi|slices` could be replaced by `use_extracted_features` to be homogeneous with classify.
- `evaluation_steps` could be a `computational resources` argument instead of optimization

### LINKED ISSUE #105: add preprocessing to mandatory arguments in generate
For random and trivial modes, preprocessing should be mandatory.
This issue will be handled as soon as a the new generate mode "shepplogan" is integrated.

### PR DESCRIPTION
**Pull request by 14thibea**
Hello,

I restored the branch in which there is the code for interpretability.
However I could not keep the commit history and everything was squashed in one commit. Is it OK for you @mdiazmel ?

New functionalities :
- `clinicadl interpret` allows to generate saliency maps linked to pretrained models based on groups of / individual images.
- `clinicadl random-search` allows to sample random networks from a predefined hyperparameter space.

New train features:
- basic data augmentation procedures (random noise, translation, blurring, erasing) are now available.
- it is possible to choose a weighted sampler instead of the default random sampler used to shuffle the training dataset.

### CODE REVIEW

**Comment by mdiazmel** [comment]
Hi @14thibea ,
It's ok. In any case, for this type of PR the better is to squash them when merging into `dev`.

**Comment by mdiazmel** [comment]
@14thibea , ready to review?

**Comment by 14thibea** [comment]
Actually, I will check that all my previous script are still working with this version.

**Comment by 14thibea** [comment]
seems ok to me !
I tested and obtained the same result than before.

**Comment by 14thibea** [comment]
Commit https://github.com/aramis-lab/AD-DL/pull/113/commits/382930521da9afa9de8d3a87c3b7a6b2c324292f closes #105 .

**Comment by mdiazmel** [comment]
Hi @14thibea , 
As discussed in the last meeting, I'll wait for the documentation associated to this PR to proceed with reviewing. If you changed your mind and prefer to submit a different PR, please let me know.

File: .gitignore [hunk 1]
```diff
@@ -35,3 +35,6 @@ clinicadl/notebooks/
 # Mask and other files
 clinicadl/clinicadl/resources/masks/*.nii
 clinicadl/clinicadl/resources/masks/*.nii.gz
+
+# Post processing specific to project
+clinicadl/post_processing
```
Comment (mdiazmel) [inline .gitignore]: What is inside this folder? Why it should be inside the clinicadl tree?

File: clinicadl/clinicadl/tools/deep_learning/models/random.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+    else:
+        if sampling_type is "exponent":
+            return 10 ** -value
+        else:
+            return value
+
+
+def random_sampling(rs_options, options):
+    """
+    Samples all the hyperparameters of the model.
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/deep_learning/models/random.py:35]: Like we are supposed to follow the Google style docstrings, can you apply the necessary changes on the docstrings of this file?

File: clinicadl/clinicadl/tools/deep_learning/models/slice_level.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from torchvision.models.resnet import BasicBlock
 from torch import nn
 import math
+from .modules import Flatten
 
 __author__ = "Junhao Wen"
```
Comment (mdiazmel) [inline clinicadl/clinicadl/tools/deep_learning/models/slice_level.py]: We can remove these lines to be coherent with the rest of the code source.

File: clinicadl/clinicadl/train/random_search.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from ..tools.deep_learning import read_json
+from ..tools.deep_learning.models.random import random_sampling
+from .train_multiCNN import train_multi_cnn
+from .train_singleCNN import train_single_cnn
+from .train_autoencoder import train_autoencoder
+
+
+def check_and_complete(rs_options):
+    """
+    This function initializes fields so a random model can be sampled.
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/random_search.py:17]: Use Google style docstrings

File: clinicadl/clinicadl/train/random_search.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        if not hasattr(rs_options, name):
+            setattr(rs_options, name, default_value)
+
+    mandatory_arguments = ['epochs', 'network_type', 'mode',
+                           'tsv_path', 'caps_dir', 'diagnoses', 'preprocessing',
+                           'n_convblocks', 'first_conv_width', 'n_fcblocks']
+
+    for argument in mandatory_arguments:
+        if not hasattr(rs_options, argument):
+            raise ValueError("The argument %s must be specified in %s." % (argument, filename))
```
Comment (mdiazmel) [inline clinicadl/clinicadl/train/random_search.py]: Try to uses python f-strings:
```suggestion
            raise ValueError(f"The argument {argument} must be specified in {filename}.")
```

File: docs/RandomSearch.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+    
+## Content of `random_search.json`
+
+`random_search.json` must be present in `launch_dir` before running the command. 
+An example of this file can be found in 
+[clinicadl/clinicadl/tools/deep_learning/models](https://github.com/aramis-lab/AD-DL/tree/master/clinicadl/clinicadl/tools/deep_learning/models).
+
+Mandatory variables:
+
+- `mode` (str) is the type of input used. Must be chosen between `image`, `patch`, `roi` and `slice`.
```
Comment (mdiazmel) [inline docs/RandomSearch.md]:
```suggestion
- `mode` (str) is the type of input used. Options are: `image`, `patch`, `roi` and `slice`. 
```

**Review by mdiazmel — COMMENTED** [review]
Thanks @14thibea ,
It's a really big and nice PR! Lots of changes!
I added some minor comments, I think that we can merge quickly and continue iterating in new PRs.

**Comment by mdiazmel** [comment]
One last thing, can you edit the PR header with the main changes that are introduced (in order to be added to the changelog for the next release). Thanks!

File: .gitignore (hunk 1 above)
Comment (14thibea) [reply .gitignore]: actually those are scripts specific to my experiments, they should not be included, I will move them elsewhere and remove this from .gitignore.

Comment (14thibea) [reply .gitignore]: Done in https://github.com/aramis-lab/AD-DL/pull/113/commits/4d4d59875977f2dd1d1c3de01f66a3e4c5d26269

File: clinicadl/clinicadl/tools/deep_learning/models/random.py (hunk 2 above)
Comment (14thibea) [reply clinicadl/clinicadl/tools/deep_learning/models/random.py:35]: Done in https://github.com/aramis-lab/AD-DL/pull/113/commits/201b3668b99a8cbb0bb13286005629b48bfbf675

File: clinicadl/clinicadl/tools/deep_learning/models/slice_level.py (hunk 3 above)
Comment (14thibea) [reply clinicadl/clinicadl/tools/deep_learning/models/slice_level.py]: Done in https://github.com/aramis-lab/AD-DL/pull/113/commits/201b3668b99a8cbb0bb13286005629b48bfbf675

File: clinicadl/clinicadl/train/random_search.py (hunk 4 above)
Comment (14thibea) [reply clinicadl/clinicadl/train/random_search.py:17]: Done in https://github.com/aramis-lab/AD-DL/pull/113/commits/201b3668b99a8cbb0bb13286005629b48bfbf675

**Review by mdiazmel — COMMENTED** [review]
Seems ok.
Let's merge this PR!
